File: OPL/C++/Source Files/Board.cpp
/**
 * @file Board.cpp
 * @brief Implementation of the Board class: covering/uncovering squares,
 *        querying board state, and finding valid square combinations.
 */

#include "../Header Files/Board.h"
#include <set>
using namespace std;

/**
 * @brief Marks a specific square as covered if it is a valid uncovered square.
 * @param square 1-based index of the square to cover
 * @return true if the square was successfully covered; false otherwise
 */
bool Board::coverSquare(const int square) {
    if (square >= 1 && square <= size && !squares[square - 1]) {
        squares[square - 1] = true;
        return true;
    }
    return false;
}

/**
 * @brief Marks a specific square as uncovered if it is currently covered.
 * @param square 1-based index of the square to uncover
 * @return true if the square was successfully uncovered; false otherwise
 */
bool Board::uncoverSquare(const int square) {
    if (square >= 1 && square <= size && squares[square - 1]) {
        squares[square - 1] = false;
        return true;
    }
    return false;
}

/**
 * @brief Checks whether a square is currently covered.
 * @param square 1-based index of the square to query
 * @return true if the specified square is covered, false otherwise
 */
bool Board::isSquareCovered(const int square) const {
    if (square >= 1 && square <= size) {
        return squares[square - 1];
    }
    return false;
}

/**
 * @brief Returns the number of squares on the board.
 * @return Board size (number of squares)
 */
int Board::getSize() const {
    return size;
}

/**
 * @brief Returns true when every square on the board is covered.
 * @return true if all squares are covered
 */
bool Board::allCovered() const {
    for (const bool square : squares) {
        if (!square) return false;
    }
    return true;
}

/**
 * @brief Returns true when every square on the board is uncovered.
 * @return true if all squares are uncovered
 */
bool Board::allUncovered() const {
    for (const bool square : squares) {
        if (square) return false;
    }
    return true;
}

/**
 * @brief Calculates the sum of all uncovered square indices for scoring.
 * @return Sum of uncovered squares
 */
int Board::getUncoveredSum() const {
    int sum = 0;
    for (int i = 1; i <= size; ++i) {
        if (!isSquareCovered(i)) {
            sum += i;
        }
    }
    return sum;
}

/**
 * @brief Calculates the sum of all covered square indices for scoring.
 * @return Sum of covered squares
 */
int Board::getCoveredSum() const {
    int sum = 0;
    for (int i = 1; i <= size; ++i) {
        if (isSquareCovered(i)) {
            sum += i;
        }
    }
    return sum;
}

/**
 * @brief Recursively finds all subsets of squares that sum to the target value.
 * @param sum Target sum to achieve from available squares
 * @param forCovering If true, consider uncovered squares to cover; otherwise consider covered squares to uncover
 * @return A set containing combinations (as sets of indices) that sum to `sum`
 */
set<set<int>> Board::findValidCombinations(const int sum, const bool forCovering) const {
    set<set<int>> combinations;

    // Iterate through all squares to find potential candidates
    for (int i = 1; i <= size; ++i) {
        // Step 1: Check if the square is available for the requested action
        if ((forCovering && !isSquareCovered(i)) || (!forCovering && isSquareCovered(i))) {

            // Step 2: Base case - exact match found
            if (i == sum) {
                combinations.insert({i});
            }
            // Step 3: Recursive case - look for remaining sum
            else if (i < sum) {
                for (set<set<int>> subCombinations = findValidCombinations(sum - i, forCovering); const set<int>& subCombination : subCombinations) {
                    // Ensure unique numbers (no repeats in combination)
                    if (!subCombination.contains(i)) {
                        set<int> combination = subCombination;
                        combination.insert(i);
                        combinations.insert(combination);
                    }
                }
            }
        }
    }
    return combinations;
}

/**
 * @brief Validates whether the provided combination is legal for the requested action.
 * @param combination Set of 1-based square indices
 * @param forCovering true when validating for covering; false for uncovering
 * @return true if every index in combination is eligible for the requested action
 */
bool Board::isValidCombination(const set<int>& combination, bool forCovering) const {
    for (const int square : combination) {
        if ((forCovering && isSquareCovered(square)) || (!forCovering && !isSquareCovered(square))) {
            return false;
        }
    }
    return true;
}

/**
 * @brief Indicates whether only a single die may be thrown under the game rule.
 * @return true when the one-die rule applies (squares starting from ONE_DIE_RULE_START are covered)
 */
bool Board::canThrowOneDie() const {
    for (int i = Board::ONE_DIE_RULE_START; i <= size; ++i) {
        if (!isSquareCovered(i)) return false;
    }
    return true;
}

File: OPL/JS/backend/models/Board.js
/**
 * Represents a game board with a set of numbered squares.
 */
class Board {
  /**
   * Creates a new Board instance.
   * @param {number} size - The number of squares on the board.
   */
  constructor(size) {
    this.size = size;
    this.squares = Array.from({ length: size }, (_, i) => i + 1);
  }

  /**
   * Covers the specified square, marking it as 0.
   * @param {number} index - The square number to cover (1-indexed).
   */
  coverSquare(index) {
    index = parseInt(index, 10);
    if (index >= 1 && index <= this.size) {
      this.squares[index - 1] = 0;
    }
  }

  /**
   * Uncovers the specified square, restoring its original number.
   * @param {number} index - The square number to uncover (1-indexed).
   */
  uncoverSquare(index) {
    if (index >= 1 && index <= this.size) {
      this.squares[index - 1] = index;
    }
  }

  /**
   * Retrieves the current values of the board squares.
   * @returns {number[]} An array representing the board squares.
   */
  getSquares() {
    return this.squares;
  }

  /**
   * Sets the board squares to the provided values.
   * @param {number[]} values - An array of values; its length must equal the board size.
   * @throws {Error} If the length of values does not match the board size.
   */
  setSquareValues(values) {
    if (values.length === this.size) {
      this.squares = values;
    } else {
      throw new Error("Invalid input: The number of values must match the board size.");
    }
  }

  /**
   * Determines if a specific square is covered.
   * @param {number} index - The square number to check (1-indexed).
   * @returns {boolean} True if the square is covered (i.e. value is 0), false otherwise.
   */
  isSquareCovered(index) {
    if (index >= 1 && index <= this.size) {
      return this.squares[index - 1] === 0;
    }
    return false;
  }

  /**
   * Checks if all squares on the board are covered.
   * @returns {boolean} True if every square is 0, false otherwise.
   */
  allCovered() {
    return this.squares.every((square) => square === 0);
  }

  /**
   * Checks if all squares on the board are uncovered.
   * @returns {boolean} True if every square is not 0, false otherwise.
   */
  allUncovered() {
    return this.squares.every((square) => square !== 0);
  }

  /**
   * Calculates the sum of indices for all covered squares.
   * @returns {number} The total sum of the positions of covered squares.
   */
  getCoveredSum() {
    return this.squares.reduce((sum, square, index) => {
      return square === 0 ? sum + (index + 1) : sum; 
    }, 0);
  }

  /**
   * Calculates the sum of indices for all uncovered squares.
   * @returns {number} The total sum of the positions of uncovered squares.
   */
  getUncoveredSum() {
    return this.squares.reduce((sum, square, index) => {
      return square !== 0 ? sum + (index + 1) : sum; 
    }, 0);
  }

  /**
   * Finds all valid combinations of square numbers that sum to a given target.
   * Takes into account board state and advantage restrictions.
   *
   * @param {number} sum - The target sum for the combination.
   * @param {boolean} forCovering - True if finding combinations for covering squares,
   *                                false if for uncovering squares.
   * @returns {number[][]} An array of valid square combinations.
   */
  findValidCombinations(sum, forCovering) {
    const combinations = [];
    const advantageSquare = this.game?.tournament?.getAdvantageSquare();
    const advantageApplied = this.game?.tournament?.getAdvantageApplied();
    const isOpponentBoard = !forCovering;
    const currentPlayer = this.game?.currentPlayer;
    const canUncover = this.game?.tournament?.canUncoverAdvantage(currentPlayer);

    /**
     * Backtracking helper to generate combinations.
     * @param {number} start - The starting square number.
     * @param {number[]} path - The current combination being built.
     * @param {number} remaining - The remaining sum required.
     */
    const backtrack = (start, path, remaining) => {
      if (remaining === 0) {
        if (this.isValidCombination(path, forCovering)) {
          if (!(advantageApplied && isOpponentBoard && path.includes(advantageSquare) && !canUncover)) {
            combinations.push([...path]);
          }
        }
        return;
      }
  
      for (let i = start; i <= this.size; i++) {
        if (advantageApplied && isOpponentBoard && i === advantageSquare && !canUncover) {
          continue;
        }
  
        if ((forCovering && !this.isSquareCovered(i)) || (!forCovering && this.isSquareCovered(i))) {
          if (i <= remaining) {
            path.push(i);
            backtrack(i + 1, path, remaining - i);
            path.pop();
          }
        }
      }
    };

    backtrack(1, [], sum);
    return combinations;
  }

  /**
   * Validates whether a given combination of squares is allowed.
   * Checks both covering logic and advantage rules.
   *
   * @param {number[]} combination - The combination of square numbers.
   * @param {boolean} forCovering - True if the combination is for covering squares,
   *                                false if for uncovering squares.
   * @returns {boolean} True if the combination is valid, false otherwise.
   */
  isValidCombination(combination, forCovering) {
    const advantageSquare = this.tournament?.getAdvantageSquare();
    const advantageApplied = this.tournament?.getAdvantageApplied();
    const isOpponentBoard = !forCovering;
    const currentPlayer = this.game?.currentPlayer;
    const canUncover = this.game?.tournament?.canUncoverAdvantage(this.game.currentPlayer);

    // Prevent uncovering the advantage square if not allowed
    if (advantageApplied && isOpponentBoard && combination.includes(advantageSquare) && !canUncover) {
      return false;
    }
    
    // For each square in the combination, ensure it is not already covered (for covering) or is covered (for uncovering)
    for (const square of combination) {
      if ((forCovering && this.isSquareCovered(square)) || (!forCovering && !this.isSquareCovered(square))) {
        return false;
      }
    }
    return true;
  }
}

module.exports = Board;